File: examples/winstt-electron/frontend/scripts/crap/score.ts
/**
 * CRAP score formula. Mirrors `examples/crap4java/src/crap4java/CrapScore.java`.
 *
 *   CRAP = CC^2 * (1 - coverage)^3 + CC
 *
 *   - `CC` is cyclomatic complexity (the function body's branch count)
 *   - `coverage` is line coverage as a fraction in [0, 1]
 *
 * Returns null when coverage is unknown (function not in the LCOV report).
 */
export function calculateCrapScore(
	complexity: number,
	coverageFraction: number | null
): number | null {
	if (coverageFraction === null) return null;
	const cc = complexity;
	const uncovered = 1 - coverageFraction;
	return cc * cc * (uncovered * uncovered * uncovered) + cc;
}

File: examples/winstt-electron/server/scripts/crap/score.py
"""CRAP score formula. Mirrors `frontend/scripts/crap/score.ts`.

    CRAP = CC^2 * (1 - coverage)^3 + CC

    - ``CC`` is cyclomatic complexity (the function body's branch count)
    - ``coverage`` is line coverage as a fraction in [0, 1]

Returns None when coverage is unknown (function not in the coverage report).
"""

from __future__ import annotations


def calculate_crap_score(complexity: int, coverage_fraction: float | None) -> float | None:
    if coverage_fraction is None:
        return None
    cc = float(complexity)
    uncovered = 1.0 - coverage_fraction
    return cc * cc * (uncovered * uncovered * uncovered) + cc
